### ISSUE TITLE
[Tkinter] [MacOS] [Python 3.10]: Setting value of ttk.Progressbar once will permanently increase CPU usage of application

### ISSUE THREAD

**Issue by fladd** [python/cpython#93037]
**Bug report**

Setting the value of a ttk.Progressbar **once**, will **permanently** increase CPU usage. The increase in CPU usage scales with the amount of widgets in the applications (yes, the other widgets seemingly unrelated to the Progressbar). In a more complex GUI, CPU usage can increase significantly (to over 90% on my system, with the fans being quite busy).

The problem can be demonstrated with the following simple test script:

```python3
import random
import tkinter as tk
import tkinter.ttk as ttk

class App(tk.Frame):
    def __init__(self, master):
        tk.Frame.__init__(self, master)
        self.grid()
        self.bar = ttk.Progressbar(self, orient="horizontal",
                                   mode='determinate', length=500, maximum=100,
                                   value=0)
        self.bar.grid(row=0, column=0)
        self.label = ttk.Label(
            self,
            text="Pressing button ONCE will permanently increase CPU usage!")
        self.label.grid(row=1, column=0)
        self.button = ttk.Button(self, text="Change Progressbar",
                                 command=self.rand_pos)
        self.button.grid(row=2, column=0)

    def rand_pos(self):
        pos = random.randint(0,100)
        self.bar["value"] = pos


if __name__ == "__main__":
    root = tk.Tk()
    root.title("Tkinter Bug")
    app = App(root)
    app.mainloop()
```

Comparing samples of the Python process [before](https://github.com/python/cpython/files/8746081/Sample.of.Python.before.txt) and several minutes [after](https://github.com/python/cpython/files/8746085/Sample.of.Python.after.txt) setting a Progressbar value (with no other interaction with the GUI), seems to indicate that Tk enters a state in which it tries to constantly update all widgets in the GUI.

NOTE: This problem does **not** occur with Python 3.9.5 on the same machine!

**Your environment**

MacBook Pro (Retina, 15-inch, Late 2013)
2 GHz Quad-Core Intel Core i7
8 GB 1600 MHz DDR3
Intel Iris Pro 1536 MB

- CPython versions tested on: Python 3.10.4 (official installer from python.org)
- Operating system and architecture: macOS Catalina 10.15.7 (19H15) x86_64

**Comment by fladd** [comment]
Further testing showed that the problem actually also occurs with Python versions 3.9.10, 3.9.11, 3.9.12 and 3.9.13. The last version that did not have this bug was 3.9.9. Notable 3.9.10 switched to Tk version 8.6.12!

**Comment by ronaldoussoren** [comment]
This is probably a bug in Tk itself, especially when the bug was introduced when switching to a newer version of Tk. 

That said, I haven't verified this with the Tk project and that should be done before closing this issue.

**Comment by fladd** [comment]
Since there seems to be neither an ActiveTcl 8.6.12 distribution nor a 8.6.12 Tcl via Homebrew, I unfortunately cannot test this directly in Tk.

Since this bug basically prevents using Progressbar widgets in any application that has a complexity (i.e. more widgets) than a simple "Hello world" demo (due to constant +90% CPU usage), is there any chance to roll back to a working Tk version (on MacOS only) for a next Python (point) release (assuming that this would pre-date an upstream fix at the Tk site)?

**Comment by ned-deily** [comment]
For these kinds of problems, the most helpful thing would be to provide a TCL version of the program which could then be tested to eliminate tkinter and Python from the issue. That said, perhaps @culler might see something obvious.

**Comment by culler** [comment]
I will take a look at this.

**Comment by culler** [comment]
I can explain much of this.  The core issue is that in the aqua theme the ttk Progressbar is animated by default.  That animation ends up triggering updates of the window whenever the animation advances.  The reason for those extra updates is complicated, and I probably don't fully understand it, and this is not the place to discuss it anyway.

The immediate fix for this problem is to disable the animation, which is pointless since on modern macs the animation does not change the appearance of a determinate progress bar.  I believe that enabling animation by default is a throwback to OSX 10.6, which used an animated barber pole for the progress bar, and that Tk was trying to emulate that.  However, there is another issue that indeterminate progress bars actually do need to be animated.  That may also be part of the reason for this default, since it is not clear to me at the moment how to have different settings for determinate and indeterminate progress bars.

The origin of the problem is contained in the following two lines
[pbarscripts.zip](https://github.com/python/cpython/files/8763982/pbarscripts.zip)
 in aquaTheme.tcl:

```
	# Enable animation for ttk::progressbar widget:
	ttk::style configure TProgressbar -period 100 -maxphase 120
```

Animation is enabled whenever the period is non-zero, so both of those numbers should be 0 on modern macs, at least for determinate progress bars.

I am attaching two modifications of the incomplete script above, one in tcl and one in python, which resolve this problem by setting the Progressbar period to 0.  These scripts also add 100 buttons to the sample window in order to make the penalty for the refresh more severe.

Please look at the python version because I think it demonstrates a bug, or at least an issue, with the ttk module.  The only way that I was able to change the default value for TProgressbar.period was to create the root window within the call to ttk.Style() and then use Tk._default_root.  Otherwise, the changes made by configuring the Style object were ignored, at least in the root window.  In Tk, all that was needed was to add the line below to the top of the script:

`ttk::style configure TProgressbar -period 0 -maxphase 0`

**Comment by fladd** [comment]
Thanks a lot for the explanation and the solution. It works in the example script I provided.

However, I realized that updating a ttk.Progressbar (i.e. setting its `value`) is still very CPU intensive in the moment on MacOS. Updating a progressbar in a loop every 100 ms will still lead to 50% CPU usage on my system. This is very different on Windows for example, where frequent updating does not have this effect on CPU usage.
